src/rgw/rgw_rados.cc

```diff
@@ -6203,27 +6205,62 @@ int RGWRados::Bucket::UpdateIndex::guard_reshard(const DoutPrefixProvider *dpp,
       ldpp_dout(dpp, 5) << "failed to get BucketShard object: ret=" << ret << dendl;
       return ret;
     }
+
     r = call(bs);
-    if (r != -ERR_BUSY_RESHARDING) {
+    if (r != -ERR_BUSY_RESHARDING && r != -ENOENT) {
```

Comment:
Could you tell me why `r != -ENOENT` check is necessary? [main branch](https://github.com/ceph/ceph/blob/8a095e903bc7518afbabe0818cc7a03d0f6cfe76/src/rgw/driver/rados/rgw_rados.cc#L6188) doesn't have this logic.

Reply:
So I dug into the code to see why -ENOENT would be returned. The value of r comes from `call(bs)`. If you look at many of the lambdas that are passed into `RGWRados::Bucket::UpdateIndex::guard_reshard`, many have had added something along the lines of `op.assert_exists()`, which means that if the bucket shard does not exist, -ENOENT will be returned.

As to why the bucket shard might not exist, that has to do with how resharding is handled prior to the significant updates between quincy and main. During resharding a new set of bucket index shards is created, data is copied into them, and if that succeeds and the bucket entry object is updated with the new bucket id, then we remove the old bucket index shards. So acknowledging the possibility that the bucket shard might no longer exist is there to handle this race.

I hope I've answered your question/concern. But please feel free to engage further! Thanks!

Reply:
@ivancich 

> If you look at many of the lambdas that are passed into RGWRados::Bucket::UpdateIndex::guard_reshard, many have had added something along the lines of op.assert_exists(), which means that if the bucket shard does not exist, -ENOENT will be returned.

I don't think so. If `op.assert_exists()` fails, it finally calls [`abort()`](https://github.com/ceph/ceph/blob/quincy/src/common/assert.cc#L75) and then the whole RGW daemon aborts.

Moreover, I suspect the following assertion does nothing in effect.

```
...
  librados::ObjectWriteOperation op;
  op.assert_exists(); // the shard must exist; if not fail rather than recreate
...
```

When `ObjectXXXOperation` is initialized, its `impl` field is set by the constructor. The succeeding `op.assert_exists()`  always succeeds because it just checks whether this field is set or not. In addition, although the comment says "the shard must exist", `op` is not associated with any shards yet.

Reply:
@satoru-takeuchi [librados::ObjectOperation::assert_exists()](https://github.com/ceph/ceph/blob/b9e2436f8/src/include/rados/librados.hpp#L350-L353) isn't a debug assertion, it's a librados API with the semantics that @ivancich describes

Reply:
Finally I understood the meaning. Thank you for your patient.